File: crates/target_csharp_system_text/output/basic_discriminator/Root.cs

using System;

using System.Text.Json;

using System.Text.Json.Serialization;

namespace JtdCodegenE2E
{
    /// <summary>

    /// </summary>

    [JsonConverter(typeof(RootJsonConverter))]
    public abstract class Root
    {
    }

    public class RootJsonConverter : JsonConverter<Root>
    {
        public override Root Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            var readerCopy = reader;
            var tagValue = JsonDocument.ParseValue(ref reader).RootElement.GetProperty("foo").GetString();

            switch (tagValue)
            {
                case "BAR_BAZ":
                    return JsonSerializer.Deserialize<RootBarBaz>(ref readerCopy, options);
                case "QUUX":
                    return JsonSerializer.Deserialize<RootQuux>(ref readerCopy, options);
                default:
                    throw new ArgumentException(String.Format("Bad Foo value: {0}", tagValue));
            }
        }

        public override void Write(Utf8JsonWriter writer, Root value, JsonSerializerOptions options)
        {
            JsonSerializer.Serialize(writer, value, value.GetType(), options);
        }
    }
}

File: crates/target_java_jackson/output/basic_discriminator/Root.java
package com.example;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "foo")
@JsonSubTypes({
    @JsonSubTypes.Type(name = "BAR_BAZ", value = RootBarBaz.class),
    @JsonSubTypes.Type(name = "QUUX", value = RootQuux.class),
})
public abstract class Root {
}
